packages/xo-server/src/xo-mixins/pool.js

```diff
@@ -0,0 +1,74 @@
+import { identity, isEmpty } from 'lodash'
+
+export default class {
+  constructor(xo) {
+    this._xo = xo
+  }
+
+  async mergeInto({ sources, target, force }) {
+    const targetHost = this._xo.getObject(target.master)
+    const _sources = sources.map(sourceId =>
+      this._xo.getObject(sourceId, 'pool')
+    )
+    let source
+    let sourceHost
+    for (source of _sources) {
+      sourceHost = this._xo.getObject(source.master)
+      if (sourceHost.productBrand !== targetHost.productBrand) {
+        throw new Error(
+          `a ${sourceHost.productBrand} pool cannot be merged into a ${
+            targetHost.productBrand
+          } pool`
+        )
+      }
+    }
+
+    const hostsPatchesDiff = {}
+    let poolPatchesDiff = []
+    for (source of _sources) {
+      poolPatchesDiff = poolPatchesDiff.concat(
+        this._xo.getPatchesDifference(source.master, target.master)
+      )
+      hostsPatchesDiff[source.id] = this._xo.getPatchesDifference(
+        target.master,
+        source.master
+      )
+    }
+
+    if (poolPatchesDiff.length > 0) {
+      const targetXapi = this._xo.getXapi(target)
+      await targetXapi.installPatches({
+        patches: await targetXapi.findPatches(poolPatchesDiff),
+      })
+    }
+
+    if (!isEmpty(hostsPatchesDiff)) {
+      let sourceXapi
+      for (source of _sources) {
+        sourceXapi = this._xo.getXapi(source)
+        await sourceXapi.installPatches({
+          patches: await sourceXapi.findPatches(hostsPatchesDiff[source.id]),
+        })
+      }
+    }
+
+    const results = []
+    let result
+    for (source of _sources) {
+      result = await this._xo
+        .mergeXenPools(source._xapiId, target._xapiId, force)
+        .catch(identity)
```

Comment:
If a merge fails, there may be an issue with the target pool and since you're merging sequentially, it may be a good idea to stop there and let the first error throw.